File: crates/hm-dsl-engine/harmont-py/harmont/_cargo.py
"""Shared cargo-argument assembly for the Rust toolchain helper.

Turns structured options (features, packages, target, …) into a
canonically-ordered, shell-safe cargo argument string. Every user-supplied
*value* (package/exclude names, the joined feature list, target triple,
profile name) is shell-quoted; the raw escape-hatch ``flags`` pass through
verbatim because they are meant to be literal cargo arguments.
"""

from __future__ import annotations

import shlex
from dataclasses import dataclass


@dataclass(frozen=True)
class CargoOpts:
    """Structured cargo invocation options. See the SDK design reference for
    the canonical token order each field lowers to."""

    workspace: bool = False
    packages: tuple[str, ...] = ()
    exclude: tuple[str, ...] = ()
    all_features: bool = False
    no_default_features: bool = False
    features: tuple[str, ...] = ()
    target: str | None = None
    all_targets: bool = False
    release: bool = False
    profile: str | None = None
    locked: bool = True
    flags: tuple[str, ...] = ()


def _validate(opts: CargoOpts) -> None:
    if opts.all_features and (opts.features or opts.no_default_features):
        msg = (
            "hm.rust: --all-features conflicts with features=/no_default_features=\n"
            f"  observed: all_features=True, features={list(opts.features)!r}, "
            f"no_default_features={opts.no_default_features!r}\n"
            "  → pass all_features=True alone, or list explicit features= "
            "without all_features"
        )
        raise ValueError(msg)
    if opts.release and opts.profile is not None:
        msg = (
            "hm.rust: release=True conflicts with profile=\n"
            f"  observed: release=True, profile={opts.profile!r}\n"
            '  → use profile="release" (identical effect) or drop one'
        )
        raise ValueError(msg)
    if opts.exclude:
        if opts.packages:
            msg = (
                "hm.rust: exclude= cannot combine with packages=\n"
                f"  observed: packages={list(opts.packages)!r}, "
                f"exclude={list(opts.exclude)!r}\n"
                "  → --exclude pairs with --workspace; packages= already selects "
                "explicitly, so drop one"
            )
            raise ValueError(msg)
        if not opts.workspace:
            msg = (
                "hm.rust: exclude= requires workspace=True\n"
                f"  observed: exclude={list(opts.exclude)!r} without workspace=True\n"
                "  → cargo --exclude only applies to --workspace; pass workspace=True"
            )
            raise ValueError(msg)


def cargo_flags(opts: CargoOpts) -> str:
    """Assemble the cargo argument middle (after the subcommand, before any
    ``--`` passthrough). Returns a leading-space string, or ``""`` when empty.
    """
    _validate(opts)
    toks: list[str] = []

    # scope
    if opts.packages:
        toks += [f"-p {shlex.quote(p)}" for p in opts.packages]
    elif opts.workspace:
        toks.append("--workspace")
        toks += [f"--exclude {shlex.quote(e)}" for e in opts.exclude]

    # target selection
    if opts.all_targets:
        toks.append("--all-targets")

    # features
    if opts.all_features:
        toks.append("--all-features")
    else:
        if opts.no_default_features:
            toks.append("--no-default-features")
        if opts.features:
            toks.append("--features " + shlex.quote(",".join(opts.features)))

    # target triple
    if opts.target is not None:
        toks.append(f"--target {shlex.quote(opts.target)}")

    # profile / release
    if opts.profile is not None:
        toks.append(f"--profile {shlex.quote(opts.profile)}")
    elif opts.release:
        toks.append("--release")

    # lockfile
    if opts.locked:
        toks.append("--locked")

    # escape hatch — verbatim
    toks += list(opts.flags)

    return (" " + " ".join(toks)) if toks else ""

File: crates/hm-dsl-engine/harmont-ts/src/toolchains/cargo.ts
// Shared cargo-argument assembly for the Rust toolchain helper.
//
// Mirrors harmont-py/harmont/_cargo.py exactly so both DSLs emit identical
// cargo command strings. User-supplied *values* are shell-quoted; raw `flags`
// pass through verbatim. `exclude` pairs with `--workspace` (cargo requires it).

export interface CargoOpts {
  readonly workspace?: boolean;
  readonly packages?: readonly string[];
  readonly exclude?: readonly string[];
  readonly allFeatures?: boolean;
  readonly noDefaultFeatures?: boolean;
  readonly features?: readonly string[];
  readonly target?: string;
  readonly allTargets?: boolean;
  readonly release?: boolean;
  readonly profile?: string;
  readonly locked?: boolean;
  readonly flags?: readonly string[];
}

// POSIX single-quote escaping, byte-for-byte identical to Python's shlex.quote:
// safe characters are left bare; everything else is wrapped in single quotes
// with embedded single quotes rendered as '"'"'. The empty string becomes ''.
const SAFE = /^[A-Za-z0-9_@%+=:,./-]+$/;
export function shQuote(s: string): string {
  if (s.length === 0) return "''";
  if (SAFE.test(s)) return s;
  return "'" + s.replace(/'/g, "'\"'\"'") + "'";
}

function validate(o: CargoOpts): void {
  if (o.allFeatures && ((o.features?.length ?? 0) > 0 || o.noDefaultFeatures)) {
    throw new Error(
      "rust: --all-features conflicts with features/noDefaultFeatures\n" +
        `  observed: allFeatures=true, features=${JSON.stringify(o.features ?? [])}, ` +
        `noDefaultFeatures=${o.noDefaultFeatures ?? false}\n` +
        "  → pass allFeatures alone, or list explicit features without allFeatures",
    );
  }
  if (o.release && o.profile !== undefined) {
    throw new Error(
      "rust: release conflicts with profile\n" +
        `  observed: release=true, profile=${JSON.stringify(o.profile)}\n` +
        '  → use profile: "release" (identical effect) or drop one',
    );
  }
  if (o.exclude?.length) {
    if (o.packages?.length) {
      throw new Error(
        "rust: exclude cannot combine with packages\n" +
          `  observed: packages=${JSON.stringify(o.packages)}, exclude=${JSON.stringify(o.exclude)}\n` +
          "  → --exclude pairs with --workspace; packages already selects explicitly, so drop one",
      );
    }
    if (!o.workspace) {
      throw new Error(
        "rust: exclude requires workspace\n" +
          `  observed: exclude=${JSON.stringify(o.exclude)} without workspace=true\n` +
          "  → cargo --exclude only applies to --workspace; pass workspace: true",
      );
    }
  }
}

export function cargoFlags(o: CargoOpts): string {
  validate(o);
  const toks: string[] = [];

  if (o.packages?.length) {
    for (const p of o.packages) toks.push(`-p ${shQuote(p)}`);
  } else if (o.workspace) {
    toks.push("--workspace");
    for (const e of o.exclude ?? []) toks.push(`--exclude ${shQuote(e)}`);
  }

  if (o.allTargets) toks.push("--all-targets");

  if (o.allFeatures) {
    toks.push("--all-features");
  } else {
    if (o.noDefaultFeatures) toks.push("--no-default-features");
    if (o.features?.length) toks.push(`--features ${shQuote(o.features.join(","))}`);
  }

  if (o.target !== undefined) toks.push(`--target ${shQuote(o.target)}`);

  if (o.profile !== undefined) toks.push(`--profile ${shQuote(o.profile)}`);
  else if (o.release) toks.push("--release");

  if (o.locked ?? true) toks.push("--locked");

  if (o.flags?.length) toks.push(...o.flags);

  return toks.length ? " " + toks.join(" ") : "";
}
